### PR TITLE
Fix post-aggregator computation when used with subtotals

### PR DESCRIPTION
**Pull request by abhishekagarwal87**
### Description

Post-aggregators, when used with subtotals, errors if the post-aggregator depends on a dimension that is absent in one of the subtotal. An example query is as follows

```
SELECT dim2, SUM(cnt), GROUPING(dim2), 
CASE WHEN GROUPING(dim2) = 1 THEN 'ALL' ELSE dim2 END
FROM druid.foo
GROUP BY GROUPING SETS ( (dim2), () )
```

The exception is 
```
java.lang.IllegalArgumentException: Missing fields [[d0]] for postAggregator [p0]

	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:148)
	at org.apache.druid.query.Queries.prepareAggregations(Queries.java:118)
	at org.apache.druid.query.groupby.GroupByQuery.<init>(GroupByQuery.java:210)
	at org.apache.druid.query.groupby.GroupByQuery.<init>(GroupByQuery.java:88)
	at org.apache.druid.query.groupby.GroupByQuery$Builder.build(GroupByQuery.java:1175)
	at org.apache.druid.query.groupby.GroupByQuery.withDimensionSpecs(GroupByQuery.java:804)
	at org.apache.druid.query.groupby.strategy.GroupByStrategyV2.processSubtotalsSpec(GroupByStrategyV2.java:446)
	at org.apache.druid.query.groupby.GroupByQueryQueryToolChest.mergeGroupByResultsWithoutPushDown(GroupByQueryQueryToolChest.java:250)
	at org.apache.druid.query.groupby.GroupByQueryQueryToolChest.mergeGroupByResults(GroupByQueryQueryToolChest.java:177)
	at org.apache.druid.query.groupby.GroupByQueryQueryToolChest.initAndMergeGroupByResults(GroupByQueryQueryToolChest.java:149)
	at org.apache.druid.query.groupby.GroupByQueryQueryToolChest.lambda$mergeResults$0(GroupByQueryQueryToolChest.java:122)
	at org.apache.druid.query.FinalizeResultsQueryRunner.run(FinalizeResultsQueryRunner.java:110)
```

I have removed the dimension renaming. We only carry over the dimensions included in the subtotal spec while generating results for any subtotal. 

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `GroupyByStrategyV2`

### CODE REVIEW

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 1]
```diff
@@ -12159,23 +12159,23 @@ public void testGroupingAggregatorWithPostAggregator() throws Exception
...
+          new Object[]{"", 3L, 0L, ""},
+          new Object[]{"a", 2L, 0L, "a"},
+          new Object[]{"abc", 1L, 0L, "abc"},
           new Object[]{NULL_STRING, 6L, 1L, "ALL"}
       );
     }
     testQuery(
         "SELECT dim2, SUM(cnt), GROUPING(dim2), \n"
-        + "CASE WHEN GROUPING(dim2) = 1 THEN 'ALL' ELSE 'INDIVIDUAL' END\n"
+        + "CASE WHEN GROUPING(dim2) = 1 THEN 'ALL' ELSE dim2 END\n"
```
Comment (gianm) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:12178]: Why did you change this test case? (As opposed to introducing a new test case.)

File: processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java [hunk 2]
```diff
@@ -442,8 +425,8 @@ public boolean doMergeResults(final GroupByQuery query)
         }
 
         GroupByQuery subtotalQuery = baseSubtotalQuery
-            .withLimitSpec(subtotalQueryLimitSpec)
-            .withDimensionSpecs(newDimensions);
+            .withLimitSpec(subtotalQueryLimitSpec);
+            //.withDimensionSpecs(newDimensions);
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:429]: Please don't include commented-out code.

File: processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java [hunk 3]
```diff
@@ -408,27 +408,10 @@ public boolean doMergeResults(final GroupByQuery query)
...
-            newDimensions.add(
-                new DefaultDimensionSpec(
-                    dimensionSpec.getOutputName(),
-                    dimensionSpec.getOutputName(),
-                    dimensionSpec.getOutputType()
-                )
-            );
             subTotalDimensionSpec.add(dimensionSpec);
-          } else {
-            // Insert dummy dimension so all subtotals queries have ResultRows with the same shape.
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:425]: Is this concern no longer valid?

IIRC, it was necessary because otherwise the ResultRows would be different lengths and so the final results wouldn't be correct.

File: processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java (hunk 2 above)
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:429]: Yup. I had removed it on my local but forgot to push.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 1 above)
Comment (abhishekagarwal87) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:12178]: I wrote this test when I submitted the patch for the grouping function. I had wanted to write it this way (as is in PR) but couldn't because of the post-aggregation bug. Now changing it as I am fixing the bug. BTW There are two more tests for the grouping function.

File: processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java (hunk 3 above)
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:425]: https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java#L581

We are still keeping all the original dimensions in the query. So result row size should be the same. I think you were concerned that the result should be null for dimensions not part of the subtotal. We are not carrying over the result for those dimensions so it should work out. 
https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java#L593

**Comment by gianm** [comment]
@abhishekagarwal87 Could you check the errors in https://travis-ci.com/github/apache/druid/jobs/456420636, including the one in GroupByQueryRunnerTest.testGroupByWithSubtotalsSpecWithLongDimensionColumn?

**Comment by abhishekagarwal87** [comment]
> @abhishekagarwal87 Could you check the errors in https://travis-ci.com/github/apache/druid/jobs/456420636, including the one in GroupByQueryRunnerTest.testGroupByWithSubtotalsSpecWithLongDimensionColumn?

It seems to be an existing bug. It can be reproduced in master by writing a subquery that generates null values for numeric dimensions.  `RowBasedGrouperHelper#getValueSuppliersForDimensions` should handle the scenario when input can be a numeric null. 

The bug occurs only when there is a subtotal or subquery used. It was caught here as when I removed the renaming of dimensions, I also removed the type change that was happening earlier (the dummy dimensions were string).

**Comment by abhishekagarwal87** [comment]
Added the test for a nested groupBy query which fails in the current master.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java [hunk 4]
```diff
@@ -712,13 +712,13 @@ public InputRawSupplierColumnSelectorStrategy makeColumnSelectorStrategy(
           return new StringInputRawSupplierColumnSelectorStrategy();
         case LONG:
           return (InputRawSupplierColumnSelectorStrategy<BaseLongColumnValueSelector>)
-              columnSelector -> columnSelector::getLong;
+              columnSelector -> () -> columnSelector.isNull() ? null : columnSelector.getLong();
```
[the remark this answers is not in the document]
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:715]: IMO, the selectors themselves should ideally cache this computation, similar to the changes being made in https://github.com/apache/druid/pull/10614. Therefore, I think this change is OK, and if there are any issues it should be fixed at the selector level.

**Review by jon-wei — APPROVED** [review]
